pyemma/coordinates/clustering/include/bits/kmeans_bits.h

```diff
@@ -0,0 +1,328 @@
+//
+// Created by marscher on 7/24/17.
+//
+
+
+#ifndef PYEMMA_KMEANS_BITS_H_H
+#define PYEMMA_KMEANS_BITS_H_H
+
+#include "kmeans.h"
+#include <pybind11/pytypes.h>
+#include <random>
+
+#undef NDEBUG
+#include <cassert>
+
+template<typename dtype>
+typename KMeans<dtype>::np_array
+KMeans<dtype>::cluster(const np_array& np_chunk, const np_array& np_centers, int n_threads) const {
+    size_t i, j;
+
+    if (np_chunk.ndim() != 2) {
+        throw std::runtime_error("Number of dimensions of \"chunk\" ain\'t 2.");
+    }
+    if (np_centers.ndim() != 2) {
+        throw std::runtime_error("Number of dimensions of \"centers\" ain\'t 2.");
+    }
+
+    size_t N_frames = np_chunk.shape(0);
+    size_t dim = np_chunk.shape(1);
+
+    if (dim == 0) {
+        throw std::invalid_argument("chunk dimension must be larger than zero.");
+    }
+
+    auto chunk = np_chunk.template unchecked<2>();
+    size_t N_centers = np_centers.shape(0);
+    auto centers = np_centers.template unchecked<2>();
+
+    std::vector<size_t> shape = {N_centers, dim};
+    py::array_t <dtype> return_new_centers(shape);
+    auto new_centers = return_new_centers.mutable_unchecked();
+    std::fill(return_new_centers.mutable_data(), return_new_centers.mutable_data() + return_new_centers.size(), 0.0);
+
+    /* initialize centers_counter and new_centers with zeros */
+    std::vector<int> centers_counter(N_centers, 0);
+
+    /* do the clustering */
+    int *centers_counter_p = centers_counter.data();
+    size_t closest_center_index = 0;
+    dtype min = std::numeric_limits<dtype>::max();
+#ifndef USE_OPENMP
+    for (i = 0; i < N_frames; i++) {
+        dtype mindist = std::numeric_limits<dtype>::max();
+        for(j = 0; j < N_centers; ++j) {
+            auto d = parent_t::metric->compute(&chunk(i, 0), &centers(j, 0));
+            if(d < mindist) {
+                mindist = d;
+                closest_center_index = j;
+            }
+        }
+        (*(centers_counter_p + closest_center_index))++;
+        for (j = 0; j < dim; j++) {
+            new_centers(closest_center_index, j) += chunk(i, j);
+        }
+    }
+#else
+    omp_set_num_threads(n_threads);
+    std::vector<dtype> dists(N_centers);
+    #pragma omp parallel
+    {
+    for (size_t i = 0; i < N_frames; ++i) {
```

Comment:
Are you sure you want to parallelize the calculation over the number of centers (which may not be much larger than the number of threads), rather than the number of frames (which is pretty much guaranteed to be huge)? As long as the parallelization fetches contiguous chunks of frames into memory during parallelization, it should still be pretty quick for each thread to do a decent chunk of work in computing the distance to all centers before moving onto the next frame.